### ISSUE TITLE
💅 `useAwait` should be type-aware and require `async` for functions that return a promise

### ISSUE THREAD

**Issue by lensbart** [biomejs/biome#1161]
### Environment information

```bash
Hello,

First of all, thanks for maintaining this project. It’s amazingly fast. I hope to one day learn some Rust so that I can contribute.

The `useAwait` that’s currently in the nursery, seems to be modeled after ESLint’s `require-await` (judging by https://github.com/biomejs/biome/issues/575). I think that in general, it’s better to mimic the behaviour of an equivalent `typescript-eslint` rule, if available, rather than the (non-type-aware) eslint rule.

In this case, for example, `useAwait` will flag functions that are marked as `async` if they don’t contain the `await` keyword, even if the function returns a promise. In constrast, [`@typescript-eslint/require-await`](https://typescript-eslint.io/rules/require-await) explicitly _adds_ `async` in this case, to inform the developer, since the return value must still be awaited.

I’d propose to follow this behaviour in the case of `useAwait` as well.

Relevant links:

- [useAwait](https://biomejs.dev/linter/rules/use-await)
- [`require-await`](https://eslint.org/docs/latest/rules/require-await)
- [`@typescript-eslint/require-await`](https://typescript-eslint.io/rules/require-await)
```

### Rule name

useAwait

### Playground link

https://biomejs.dev/playground/?lintRules=all&code=YQBzAHkAbgBjACAAZgB1AG4AYwB0AGkAbwBuACAAcgBlAHQAdQByAG4AcwBQAHIAbwBtAGkAcwBlADEAKAApACAAewAKACAAIAByAGUAdAB1AHIAbgAgAFAAcgBvAG0AaQBzAGUALgByAGUAcwBvAGwAdgBlACgAMQApADsACgB9AAoACgBjAG8AbgBzAHQAIABfAHIAZQB0AHUAcgBuAHMAUAByAG8AbQBpAHMAZQAyACAAPQAgACgAKQAgAD0APgAgAHIAZQB0AHUAcgBuAHMAUAByAG8AbQBpAHMAZQAxACgAKQA7AAoACgA%3D

### Expected result

No error for functions marked as `async` that don’t contain `await` but do return a promise

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
@lensbart Biome doesn't have the infrastructure yet to deduce types, so it's impossible at the moment to match the rule from TypeScript Eslint.

Unless there's an issue with the **current logic of the rule**, I advise to close the issue.

We will soon publish more info about this topic.

**Comment by mctrafik** [comment]
+1 this feature request.

The biggest reason for it in my company that eslint is slow is because it does type inference and we only need it for this one rule: `https://typescript-eslint.io/rules/no-floating-promises/`. Without it, every few months someone introduces a race condition and we spend a week debugging it. Blocker for us using from fully switching to biome.

**Comment by john-griffin** [comment]
Yeah we would really like `no-floating-promises` too.

**Comment by jpenna** [comment]
I'm going to eslint because https://typescript-eslint.io/rules/no-floating-promises/ is missing too, unfortunately :(
